homie-controller/tests/integration.rs

```diff
@@ -0,0 +1,226 @@
+//! Integration test between homie-controller and homie-device crates. Starts a device and a
+//! controller connected to the same MQTT broker, and ensures that the controller can discover the
+//! device as expected.
+
+use futures::future::ready;
+use homie_controller::{Event, HomieController, State};
+use homie_device::{HomieDevice, Node, Property, SpawnError};
+use librumqttd::{Broker, Config};
+use rumqttc::{ConnectionError, MqttOptions, StateError};
+use std::io::ErrorKind;
+use std::sync::mpsc;
+use std::thread;
+
+// A high port number which is hopefully not in use, to use for the MQTT broker.
+const PORT: u16 = 10883;
+
+#[tokio::test]
+async fn test_device() {
+    // Start MQTT broker.
+    spawn_mqtt_broker(PORT);
+
+    // Start controller.
+    let controller_options = MqttOptions::new("homie_controller", "localhost", PORT);
+    let (controller, mut event_loop) = HomieController::new(controller_options, "homie");
+    controller.start().await.unwrap();
+
+    // Start device
+    let (updates_tx, updates_rx) = mpsc::sync_channel(10);
+    let device_options = MqttOptions::new("homie_device", "localhost", PORT);
+    let mut device_builder = HomieDevice::builder("homie/device_id", "Device name", device_options);
+    device_builder.set_update_callback(move |node_id, property_id, value| {
+        assert_eq!(property_id, "property_id");
+        assert_eq!(node_id, "node_id");
+        updates_tx.send(value.clone()).unwrap();
+        ready(Some(value))
+    });
+    let (mut homie, homie_handle) = device_builder.spawn().await.unwrap();
```

Comment:
I guess you could also turn this into: 
```suggestion
    let (mut homie, homie_handle) = device_builder.spawn().await.expect("this may fail if port 10883 is already in use");
```

Reply:
I would expect it to be the `broker.start()` that fails, so I'll add an expect there.